lib/rack/response.rb

```diff
@@ -110,14 +111,14 @@ def finish(&block)
         close
         return [@status, @headers, []]
       else
-        if @length && @length > 0 && !chunked?
-          set_header CONTENT_LENGTH, @length.to_s
-        end
-
         if block_given?
           @block = block
           return [@status, @headers, self]
         else
+          # If we know the length of the body, set the content-length header... except if we are chunked? which is a legacy special case where the body might already be encoded and thus the actual length and the content-length are likely to be different.
```

Comment:
I think the reason we aren't potentially setting the content-length if a block is given is because the block could write to the body, and that would result in an invalid content length? If that is correct, can you add a comment explaining that?

Reply:
Yes, that's correct, sure I can explain that too.